cktap/_ecdsa.py

```diff
@@ -0,0 +1,302 @@
+    if len(pub) == 65 and pub[0] == 4:
+        return 'bin'
+    elif len(pub) == 33 and pub[0] in [2, 3]:
+        return 'bin_compressed'
+    else:
+        raise Exception("Pubkey not in recognized format")
+
+
+def encode_pubkey(pub, formt):
+    if not isinstance(pub, (tuple, list)):
+        pub = decode_pubkey(pub)
+    if formt == 'decimal':
+        return pub
+    elif formt == 'bin':
+        return b'\x04' + encode_base256(pub[0], 32) + encode_base256(pub[1], 32)
+    elif formt == 'bin_compressed':
+        return bytes([2 + (pub[1] % 2)]) + encode_base256(pub[0], 32)
+    raise ValueError("Invalid format!")
+
+
+def decode_privkey(priv, formt=None):
+    if formt == 'decimal':
+        return priv
+    elif formt == 'bin':
+        return decode_base256(priv)
+    elif formt == 'bin_compressed':
+        return decode_base256(priv[:32])
+    raise ValueError("Invalid format!")
+
+
+def decode_pubkey(pub, formt=None):
+    """takes pubkey, detects type, returns tuple of (x, y)"""
+    if not formt:
+        formt = get_pubkey_format(pub)
+    if formt == 'decimal':
+        return pub
+    elif formt == 'bin':
+        return decode_base256(pub[1:33]), decode_base256(pub[33:65])
+    elif formt == 'bin_compressed':
+        x = decode_base256(pub[1:33])
+        beta = pow(int(x * x * x + A * x + B), int((P + 1) // 4), int(P))
+        y = (P - beta) if ((beta + pub[0])) % 2 else beta
+        return (x, y)
+    else:
+        raise Exception("Invalid format!")
+
+
+def privkey_to_pubkey(privkey: bytes):
+    privkey = decode_base256(privkey)
+    if privkey >= N:
+        raise ValueError("Invalid privkey")
+    return encode_pubkey(fast_multiply(G, privkey), "bin_compressed")
+
+
+# EDCSA
+
+
+def encode_sig(v, r, s) -> bytes:
+    vb, rb, sb = bytes([v]), encode_base256(r, 32), encode_base256(s, 32)
+    return vb + rb + sb
+
+
+def decode_sig(sig: bytes):
+    return sig[0], decode_base256(sig[1:33]), decode_base256(sig[33:])
+
+
+# https://tools.ietf.org/html/rfc6979#section-3.2
+def deterministic_generate_k(msghash, priv):
+    hmac_sha256 = lambda k, s: hmac.new(k, s, hashlib.sha256)
+    v = bytearray(b'\1' * 32)
+    k = bytearray(32)  # b'\0' * 32
+    msghash = encode_base256(decode_base256(msghash), 32)  # encode msg hash as 32 bytes
+    k = hmac_sha256(k, v + b'\0' + priv + msghash).digest()
+    v = hmac_sha256(k, v).digest()
+    k = hmac_sha256(k, v + b'\1' + priv + msghash).digest()
+    v = hmac_sha256(k, v).digest()
+    res = hmac_sha256(k, v).digest()
+    return decode_base256(res)
+
+
+# MSG SIGNING
+
+def ecdsa_raw_sign(msghash, priv):
+    """sign msg hash (z) with privkey & RFC6979 (k);
+    returns signature (v,r,s) with low s (BIP66) by default"""
+    z = decode_base256(msghash)
+    k = deterministic_generate_k(msghash, priv)
+
+    r, y = fast_multiply(G, k)
+    s = inv(k, N) * (z + r * decode_base256(priv)) % N
+
+    is_compressed = True  # TODO do we plan to support uncompressed?
```

Comment:
we do not support long private keys only 32bytes - so this is imo good

Reply:
right compressed keys only, everywhere